kv/memberlist/dnsprovider.go

```diff
@@ -13,3 +15,31 @@ type DNSProvider interface {
 	// Addresses returns the latest addresses present in the DNSProvider.
 	Addresses() []string
 }
+
+type dnsProvider struct {
+	sync.Mutex
+	addr []string
+}
+
+func NewDNSProvider() DNSProvider { return &dnsProvider{} }
+
+func (d *dnsProvider) Resolve(ctx context.Context, addrs []string) error {
+	d.Lock()
+	defer d.Unlock()
+	for _, a := range addrs {
+		ips, err := net.LookupIP(a)
+		if err != nil {
+			return err
+		}
+		for _, ip := range ips {
+			d.addr = append(d.addr, ip.String())
```

Comment:
Instead of doing an `append(d.addr)` in a loop, I'd rather see a new slice of addresses constructed and then swapped at the end. That prevents cases where an error is returned and `d.addr` is only half updated. This would also allow you to avoid holding the lock while doing network operations. Something like:

```go
var addrs []string
for _, a: range addrs {
    // do the lookup
    addrs = append(addrs, ip.String())
}
d.Lock()
d.addrs = addrs
d.Unlock()
return nil
```